File: latest/week1/workshop2_m6/W02_M06_demo.R
# ============================================================
# Workshop 2 — Module 6: Themes & Publication-Quality Styling
# R Demonstration Script — UNYT Tirana
# ============================================================
library(tidyverse)
library(patchwork)
dir.create("output", showWarnings = FALSE)

# ── 1. FIVE BUILT-IN THEMES ─────────────────────────────
p_base <- ggplot(mpg, aes(x = displ, y = hwy)) + geom_point(alpha = 0.5, size = 1.5)

p_themes <- (
  (p_base + theme_gray() + labs(title = "theme_gray()")) |
  (p_base + theme_minimal() + labs(title = "theme_minimal()")) |
  (p_base + theme_classic() + labs(title = "theme_classic()"))
) / (
  (p_base + theme_bw() + labs(title = "theme_bw()")) |
  (p_base + theme_void() + labs(title = "theme_void()")) |
  (p_base + theme_dark() + labs(title = "theme_dark()"))
)
ggsave("output/five_themes.png", p_themes, width = 12, height = 7, dpi = 300)

# ── 2. BEFORE vs AFTER ──────────────────────────────────
df <- mpg |> count(class, sort = TRUE) |> mutate(class = fct_reorder(class, n))

p_before <- ggplot(df, aes(x = class, y = n)) +
  geom_col(fill = "steelblue") +
  coord_flip() +
  labs(title = "BEFORE: Defaults")

p_after <- ggplot(df, aes(x = class, y = n, fill = class == "suv")) +
  geom_col(width = 0.5, show.legend = FALSE) +
  scale_fill_manual(values = c("FALSE" = "#1565C0", "TRUE" = "#E53935")) +
  geom_text(aes(label = n), hjust = -0.2, size = 3.5, fontface = "bold") +
  coord_flip() +
  theme_minimal(base_size = 11) +
  theme(
    plot.title = element_text(face = "bold", size = 14),
    panel.grid.major.y = element_blank(),
    panel.grid.minor = element_blank(),
    axis.title = element_blank()
  ) +
  labs(title = "AFTER: Custom theme")

ggsave("output/before_after.png", p_before + p_after, width = 10, height = 5, dpi = 300)

# ── 3. CUSTOM THEME FUNCTION ────────────────────────────
theme_unyt <- function(base_size = 11) {
  theme_minimal(base_size = base_size) +
    theme(
      plot.title = element_text(face = "bold", size = rel(1.3), margin = margin(b = 8)),
      plot.subtitle = element_text(color = "grey50", margin = margin(b = 12)),
      plot.caption = element_text(face = "italic", size = rel(0.8), hjust = 0, color = "grey60"),
      panel.grid.major.x = element_blank(),
      panel.grid.minor = element_blank(),
      axis.title = element_text(face = "bold"),
      legend.position = "bottom",
      legend.title = element_text(face = "bold"),
      plot.margin = margin(t = 10, r = 15, b = 10, l = 10)
    )
}

p_branded <- ggplot(mpg, aes(x = displ, y = hwy, color = factor(cyl))) +
  geom_point(alpha = 0.6, size = 2) +
  geom_smooth(method = "lm", se = FALSE, linewidth = 0.8) +
  scale_color_brewer(palette = "Set2") +
  labs(
    title = "Engine Size vs Fuel Economy",
    subtitle = "Coloured by cylinder count, with per-group trend lines",
    caption = "Source: mpg dataset | UNYT Data Visualization Course",
    x = "Displacement (L)", y = "Highway MPG", color = "Cylinders"
  ) +
  theme_unyt()

ggsave("output/branded_theme.png", p_branded, width = 7, height = 5, dpi = 300)
ggsave("output/branded_theme.pdf", p_branded, width = 7, height = 5, device = cairo_pdf)

# ── 4. GGTHEMES ──────────────────────────────────────────
# install.packages("ggthemes")
# library(ggthemes)
# p_base + theme_economist() + scale_color_economist()
# p_base + theme_wsj()
# p_base + theme_fivethirtyeight()
# p_base + theme_tufte()

# ── 5. EXPORT DEMO ──────────────────────────────────────
ggsave("output/export_pdf.pdf", p_branded, width = 7, height = 5, device = cairo_pdf)
ggsave("output/export_png_300.png", p_branded, width = 7, height = 5, dpi = 300)
ggsave("output/export_png_150.png", p_branded, width = 7, height = 5, dpi = 150)
ggsave("output/export_svg.svg", p_branded, width = 7, height = 5)

cat("\n── All W02-M06 R plots saved ──\n")
cat("Compare file sizes:\n")
cat(system("ls -lh output/export_*", intern = TRUE), sep = "\n")

File: latest/week1/workshop2_m6/W02_M06_demo.py
"""W02-M06: Themes & Publication-Quality Styling — Python — UNYT Tirana"""
import numpy as np, matplotlib.pyplot as plt, os
os.makedirs("output", exist_ok=True); np.random.seed(42)

# 1. GLOBAL THEME via rcParams
plt.rcParams.update({
    "figure.figsize": (7, 5), "figure.dpi": 150, "font.size": 11,
    "axes.spines.top": False, "axes.spines.right": False,
    "axes.grid": True, "grid.alpha": 0.12, "axes.facecolor": "white",
    "axes.titleweight": "bold", "axes.labelweight": "bold",
})

# 2. BEFORE vs AFTER
np.random.seed(22); x=np.random.uniform(1,6,60); y=0.5*x+np.random.normal(0,0.8,60)

fig,(ax1,ax2)=plt.subplots(1,2,figsize=(10,4))
# Before: raw matplotlib defaults (temporarily reset)
ax1.scatter(x,y,s=20,c="steelblue")
ax1.set_title("BEFORE: Defaults",color="#C62828")
ax1.set_facecolor("#EBEBEB"); ax1.grid(True,color="white")
for sp in ax1.spines.values(): sp.set_visible(True)
# After: rcParams applied
ax2.scatter(x,y,s=25,c="#1565C0",alpha=0.6,edgecolors="white",linewidth=0.3)
ax2.set_title("AFTER: rcParams + manual")
fig.suptitle("Theme Transformation",fontsize=12,fontweight="bold")
plt.tight_layout(); plt.savefig("output/before_after_py.png",dpi=300); plt.close()

# 3. REUSABLE THEME FUNCTION
def apply_unyt_theme(ax, title="", subtitle="", caption=""):
    """Apply UNYT branded styling to any axes."""
    ax.spines["top"].set_visible(False)
    ax.spines["right"].set_visible(False)
    ax.set_facecolor("white")
    ax.grid(axis="y", alpha=0.12)
    ax.tick_params(labelsize=9)
    if title: ax.set_title(title, fontsize=13, fontweight="bold", pad=10)
    if caption:
        ax.figure.text(0.12, 0.01, caption, fontsize=7, color="#AAA", fontstyle="italic")

np.random.seed(22); n=80; x2=np.random.uniform(1,6,n); y2=0.5*x2+np.random.normal(0,0.8,n)
cls=np.random.choice(["A","B","C"],n)
pal={"A":"#1565C0","B":"#E53935","C":"#2E7D32"}

fig,ax=plt.subplots(figsize=(7,5))
for c,col in pal.items():
    mask=cls==c; ax.scatter(x2[mask],y2[mask],s=30,c=col,alpha=0.6,edgecolors="white",linewidth=0.4,label=c)
ax.legend(fontsize=8,title="Class",title_fontsize=9,frameon=False)
ax.set_xlabel("Displacement (L)"); ax.set_ylabel("Highway MPG")
apply_unyt_theme(ax, title="Engine Size vs Fuel Economy",
                 caption="Source: mpg dataset | UNYT Data Visualization Course")
plt.tight_layout()
plt.savefig("output/branded_py.pdf", bbox_inches="tight")
plt.savefig("output/branded_py.png", dpi=300, bbox_inches="tight")
plt.close()

# 4. plt.style.use() demo
styles = ["seaborn-v0_8-whitegrid", "ggplot", "bmh", "fivethirtyeight"]
fig, axes = plt.subplots(1, 4, figsize=(14, 3.5))
for ax, s in zip(axes, styles):
    with plt.style.context(s):
        ax.scatter(x, y, s=15, alpha=0.6)
        ax.set_title(s, fontsize=7, fontweight="bold")
        ax.tick_params(labelsize=5)
fig.suptitle("plt.style.use(): Built-In Style Sheets", fontsize=11, fontweight="bold")
plt.tight_layout(); plt.savefig("output/styles_py.png", dpi=300); plt.close()

# 5. EXPORT comparison
fig, ax = plt.subplots(figsize=(7, 5))
ax.scatter(x2, y2, s=20, c="#1565C0", alpha=0.6)
apply_unyt_theme(ax, title="Export Test")
plt.tight_layout()
fig.savefig("output/export_test.pdf", bbox_inches="tight")
fig.savefig("output/export_test.png", dpi=300, bbox_inches="tight")
fig.savefig("output/export_test.svg", bbox_inches="tight")
plt.close()

print("All W02-M06 Python plots saved")
for f in ["export_test.pdf", "export_test.png", "export_test.svg"]:
    sz = os.path.getsize(f"output/{f}")
    print(f"  {f}: {sz/1024:.0f} KB")
